### PR TITLE
Validate Maven profile names and recognize -P

### PR DESCRIPTION
**Pull request by aloubyansky**
Fix for #2043

### CODE REVIEW

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java [hunk 1]
```diff
@@ -197,18 +197,47 @@ public static DefaultRepositorySystemSession newSession(RepositorySystem system,
...
     public static List<RemoteRepository> getRemoteRepos(Settings settings) throws AppModelResolverException {
         final List<RemoteRepository> remotes = new ArrayList<>();
+        // first it's the active by default ones
         for (Profile profile : settings.getProfiles()) {
             if (profile.getActivation() != null && profile.getActivation().isActiveByDefault()) {
                 addProfileRepos(profile, remotes);
             }
         }
+        // then it's  the ones activated with -P
+        if(mvnCmd != null) {
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java]: Hmmm. That seems very fragile. There's no other way to do that?

BTW, Maven supports multiple `-P` options.

Comment (aloubyansky) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java]: > Hmmm. That seems very fragile. There's no other way to do that?

I couldn't find any alternative so far.
Imagine you are writing a JUnit test and in the test you need to figure out the Maven config that launched that test (also it's possible that the test could have been launched not from Maven).
That essentially means that all that's available for us to discover is system props and env. There is also classpath and filesystem though but not sure that could help much?

> BTW, Maven supports multiple `-P` options.

Thanks! Added support for multiple `-P`'s,

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java [hunk 2]
```diff
@@ -68,15 +67,17 @@
     private static final String MAVEN_HOME = "maven.home";
     private static final String M2_HOME = "M2_HOME";
     private static final String SETTINGS_XML = "settings.xml";
+    private static final String DASH_P = "-P";
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java]: @aloubyansky wondering if you should scan for ` -P` instead, with a space. That would be a bit less fragile.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java [hunk 3]
```diff
@@ -197,18 +198,56 @@ public static DefaultRepositorySystemSession newSession(RepositorySystem system,
...
                 addProfileRepos(profile, remotes);
             }
         }
+        // then it's  the ones activated with -P
+        if(mvnCmd != null) {
+            final int cmdLength = mvnCmd.length();
+            StringBuilder buf = null;
+            int i = mvnCmd.indexOf(DASH_P);
+            while(i >= 0) {
+                i += 2;
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java]: You should use `DASH_P.length()`.

**Review by gsmet — CHANGES_REQUESTED** [review]
@aloubyansky added 2 minor comments. Could you take care of them?

Thanks!

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java (hunk 2 above)
Comment (aloubyansky) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java]: Then -PprofileName won't work.

**Comment by aloubyansky** [comment]
The current edition supports:
* -P arg[,arg]+
* --activate-profiles arg[,arg]+
* -P=arg[,arg]+ // looking bizarre but maven does recognize that
* --activate-profiles=arg[,arg]+
* multiple occurrences of -P and --activate-profiles on the same command line
* mix of -P and --activate-profiles on the same line

If the same profile is activated and deactivated at the same time, it's effectively deactivated which is consistent with the maven behavior.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java (hunk 2 above)
Comment (gsmet) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java]: @aloubyansky I'm still a bit worried about the fragility of this. I'm wondering if we could use the same machinery that Maven uses to parse the options? I don't know if it's something accessible and usable?

Comment (aloubyansky) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java]: I re-wrote it using API from maven-embedder (which is a new dependency).

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java [hunk 4]
```diff
@@ -197,18 +221,49 @@ public static DefaultRepositorySystemSession newSession(RepositorySystem system,
...
+            }
+        }
+
         for (Profile profile : settings.getProfiles()) {
-            if (profile.getActivation() != null && profile.getActivation().isActiveByDefault()) {
+            if(inactiveProfiles.contains(profile.getId())) {
+                continue;
+            }
+            if (activeProfiles.contains(profile.getId())
+                    || profile.getActivation() != null && profile.getActivation().isActiveByDefault()) {
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java]: any chance you could add parentheses here for the second grouped condition? Not that it's incorrect but it's better to be extra clear about it.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java [hunk 5]
```diff
@@ -218,6 +273,14 @@ public static DefaultRepositorySystemSession newSession(RepositorySystem system,
         return remotes;
     }
 
+    private static Profile getProfile(String name, Settings settings) throws AppModelResolverException {
+        final Profile profile = settings.getProfilesAsMap().get(name);
+        if(profile == null) {
+            throw new AppModelResolverException("Failed to locate profile " + name + " in Maven settings");
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java]: I think Maven itself logs a warning when you have an unrecognized profile. Isn't throwing an exception a bit too harsh?

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks, I really like this approach better. I had 2 minor comments though. Could you take a look? The first one requires an action, not sure about the second one (the exception vs warn thing).

**Review by gsmet — APPROVED** [review]

**Pull request merged**